File: app/src/test/java/org/alkaline/taskbrain/data/NoteReconstructionTest.kt
package org.alkaline.taskbrain.data

import org.junit.Assert.*
import org.junit.Test

class NoteReconstructionTest {

    private fun note(
        id: String,
        content: String = "",
        containedNotes: List<String> = emptyList(),
        rootNoteId: String? = null,
        parentNoteId: String? = null,
        state: String? = null,
    ) = Note(
        id = id,
        content = content,
        containedNotes = containedNotes,
        rootNoteId = rootNoteId,
        parentNoteId = parentNoteId,
        state = state,
    )

    // --- rebuildAllNotes ---

    @Test
    fun `rebuildAll - single root with no children`() {
        val rawNotes = mapOf("a" to note("a", "Alpha"))
        val result = rebuildAllNotes(rawNotes)
        assertEquals(1, result.notes.size)
        assertEquals("Alpha", result.notes[0].content)
        assertTrue(result.notesNeedingFix.isEmpty())
    }

    @Test
    fun `rebuildAll - root with descendants reconstructs content`() {
        val root = note("r", "Root", containedNotes = listOf("c1", "c2"))
        val c1 = note("c1", "Child 1", rootNoteId = "r", parentNoteId = "r")
        val c2 = note("c2", "Child 2", rootNoteId = "r", parentNoteId = "r")
        val rawNotes = mapOf("r" to root, "c1" to c1, "c2" to c2)

        val result = rebuildAllNotes(rawNotes)
        assertEquals(1, result.notes.size)
        assertEquals("Root\nChild 1\nChild 2", result.notes[0].content)
        assertTrue(result.notesNeedingFix.isEmpty())
    }

    @Test
    fun `rebuildAll - multiple roots`() {
        val rawNotes = mapOf(
            "a" to note("a", "Alpha"),
            "b" to note("b", "Beta"),
        )
        val result = rebuildAllNotes(rawNotes)
        assertEquals(2, result.notes.size)
        val ids = result.notes.map { it.id }.toSet()
        assertTrue(ids.contains("a"))
        assertTrue(ids.contains("b"))
    }

    @Test
    fun `rebuildAll - orphaned child excluded from top-level`() {
        val rawNotes = mapOf(
            "c1" to note("c1", "Orphan", rootNoteId = "missing", parentNoteId = "missing"),
        )
        val result = rebuildAllNotes(rawNotes)
        assertEquals(0, result.notes.size)
    }

    @Test
    fun `rebuildAll - deleted root still returned`() {
        val rawNotes = mapOf(
            "a" to note("a", "Active"),
            "b" to note("b", "Deleted", state = "deleted"),
        )
        val result = rebuildAllNotes(rawNotes)
        assertEquals(2, result.notes.size)
    }

    @Test
    fun `rebuildAll - empty map`() {
        val result = rebuildAllNotes(emptyMap())
        assertTrue(result.notes.isEmpty())
        assertTrue(result.notesNeedingFix.isEmpty())
    }

    @Test
    fun `rebuildAll - orphan refs in containedNotes are dropped and mark needsFix`() {
        // Parent refers to c1, c2, c3, but only c2 exists as a real child
        val root = note("r", "Root", containedNotes = listOf("c1", "c2", "c3"))
        val c2 = note("c2", "Only child", rootNoteId = "r", parentNoteId = "r")
        val rawNotes = mapOf("r" to root, "c2" to c2)

        val result = rebuildAllNotes(rawNotes)
        assertEquals(1, result.notes.size)
        assertEquals("Root\nOnly child", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    @Test
    fun `rebuildAll - stray child linked via parentNoteId appended at end and marks needsFix`() {
        // Parent doesn't know about c2, but c2 has parentNoteId = root
        val root = note("r", "Root", containedNotes = listOf("c1"))
        val c1 = note("c1", "Known", rootNoteId = "r", parentNoteId = "r")
        val c2 = note("c2", "Stray", rootNoteId = "r", parentNoteId = "r")
        val rawNotes = mapOf("r" to root, "c1" to c1, "c2" to c2)

        val result = rebuildAllNotes(rawNotes)
        assertEquals("Root\nKnown\nStray", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    @Test
    fun `rebuildAll - all declared children missing but strays present reconstructs from strays`() {
        // This is the user's scenario: containedNotes all orphaned, but parentNoteId links exist
        val root = note("r", "Alarms (dupe)", containedNotes = listOf("orphan1", "orphan2"))
        val real1 = note("real1", "Real one", parentNoteId = "r")
        val real2 = note("real2", "Real two", parentNoteId = "r")
        val rawNotes = mapOf("r" to root, "real1" to real1, "real2" to real2)

        val result = rebuildAllNotes(rawNotes)
        assertEquals("Alarms (dupe)\nReal one\nReal two", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    @Test
    fun `rebuildAll - deleted child dropped and marks needsFix`() {
        val root = note("r", "Root", containedNotes = listOf("c1", "c2"))
        val c1 = note("c1", "Live", parentNoteId = "r")
        val c2 = note("c2", "Deleted", parentNoteId = "r", state = "deleted")
        val rawNotes = mapOf("r" to root, "c1" to c1, "c2" to c2)

        val result = rebuildAllNotes(rawNotes)
        assertEquals("Root\nLive", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    @Test
    fun `rebuildAll - child with mismatched parentNoteId dropped from ref list`() {
        // Parent claims c1, but c1's parentNoteId points elsewhere
        val root = note("r", "Root", containedNotes = listOf("c1"))
        val c1 = note("c1", "Not mine", parentNoteId = "someone-else")
        val rawNotes = mapOf("r" to root, "c1" to c1)

        val result = rebuildAllNotes(rawNotes)
        assertEquals("Root", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    @Test
    fun `rebuildAll - nested strays render at correct depth`() {
        val root = note("r", "Root", containedNotes = listOf("c1"))
        val c1 = note("c1", "Known", containedNotes = listOf(), parentNoteId = "r")
        val grandchild = note("gc1", "Stray gc", parentNoteId = "c1")
        val rawNotes = mapOf("r" to root, "c1" to c1, "gc1" to grandchild)

        val result = rebuildAllNotes(rawNotes)
        assertEquals("Root\nKnown\n\tStray gc", result.notes[0].content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    // --- rebuildAffectedNotes ---

    @Test
    fun `rebuildAffected - single root changed`() {
        val existing = listOf(note("a", "Old"), note("b", "Beta"))
        val rawNotes = mapOf(
            "a" to note("a", "New"),
            "b" to note("b", "Beta"),
        )
        val result = rebuildAffectedNotes(existing, setOf("a"), rawNotes)
        assertEquals(2, result.notes.size)
        assertEquals("New", result.notes.find { it.id == "a" }!!.content)
        assertEquals("Beta", result.notes.find { it.id == "b" }!!.content)
    }

    @Test
    fun `rebuildAffected - root deleted`() {
        val existing = listOf(note("a", "Alpha"), note("b", "Beta"))
        val rawNotes = mapOf("b" to note("b", "Beta"))
        val result = rebuildAffectedNotes(existing, setOf("a"), rawNotes)
        assertEquals(1, result.notes.size)
        assertEquals("b", result.notes[0].id)
    }

    @Test
    fun `rebuildAffected - new root added`() {
        val existing = listOf(note("a", "Alpha"))
        val rawNotes = mapOf(
            "a" to note("a", "Alpha"),
            "b" to note("b", "Beta"),
        )
        val result = rebuildAffectedNotes(existing, setOf("b"), rawNotes)
        assertEquals(2, result.notes.size)
    }

    @Test
    fun `rebuildAffected - unaffected preserved by reference`() {
        val noteA = note("a", "Alpha")
        val noteB = note("b", "Beta")
        val existing = listOf(noteA, noteB)
        val rawNotes = mapOf(
            "a" to note("a", "Changed"),
            "b" to noteB,
        )
        val result = rebuildAffectedNotes(existing, setOf("a"), rawNotes)
        assertSame(noteB, result.notes.find { it.id == "b" })
    }

    @Test
    fun `rebuildAffected - no changes returns same list reference`() {
        val existing = listOf(note("a", "Alpha"))
        val rawNotes = mapOf("a" to note("a", "Alpha"))
        val result = rebuildAffectedNotes(existing, setOf("nonexistent"), rawNotes)
        assertSame(existing, result.notes)
    }

    @Test
    fun `rebuildAffected - note becomes child removed from top-level`() {
        val existing = listOf(note("a", "Alpha"), note("b", "Beta"))
        val rawNotes = mapOf(
            "a" to note("a", "Alpha"),
            "b" to note("b", "Beta", parentNoteId = "a"),
        )
        val result = rebuildAffectedNotes(existing, setOf("b"), rawNotes)
        assertEquals(1, result.notes.size)
        assertEquals("a", result.notes[0].id)
    }

    @Test
    fun `rebuildAffected - fix applied marks notesNeedingFix`() {
        val existing = listOf(note("r", "Old"))
        val root = note("r", "New", containedNotes = listOf("orphan"))
        val rawNotes = mapOf("r" to root)
        val result = rebuildAffectedNotes(existing, setOf("r"), rawNotes)
        assertEquals("New", result.notes.find { it.id == "r" }!!.content)
        assertEquals(setOf("r"), result.notesNeedingFix)
    }

    // --- reconstructNoteContent ---

    private fun childrenByParent(rawNotes: Map<String, Note>): Map<String, List<Note>> {
        val result = mutableMapOf<String, MutableList<Note>>()
        for (n in rawNotes.values) {
            val p = n.parentNoteId ?: continue
            if (n.state == "deleted") continue
            result.getOrPut(p) { mutableListOf() }.add(n)
        }
        return result
    }

    @Test
    fun `reconstructContent - no children or refs returns same instance`() {
        val n = note("a", "Simple note")
        val raw = mapOf("a" to n)
        val (result, fixed) = reconstructNoteContent(n, raw, childrenByParent(raw))
        assertSame(n, result)
        assertFalse(fixed)
    }

    @Test
    fun `reconstructContent - declared children in order`() {
        val root = note("r", "Root", containedNotes = listOf("c1", "c2"))
        val c1 = note("c1", "First", parentNoteId = "r")
        val c2 = note("c2", "Second", parentNoteId = "r")
        val raw = mapOf("r" to root, "c1" to c1, "c2" to c2)
        val (result, fixed) = reconstructNoteContent(root, raw, childrenByParent(raw))
        assertEquals("Root\nFirst\nSecond", result.content)
        assertFalse(fixed)
    }

    @Test
    fun `reconstructContent - spacer in containedNotes renders as blank line`() {
        val root = note("r", "Root", containedNotes = listOf("c1", "", "c2"))
        val c1 = note("c1", "A", parentNoteId = "r")
        val c2 = note("c2", "B", parentNoteId = "r")
        val raw = mapOf("r" to root, "c1" to c1, "c2" to c2)
        val (result, fixed) = reconstructNoteContent(root, raw, childrenByParent(raw))
        assertEquals("Root\nA\n\nB", result.content)
        assertFalse(fixed)
    }

    @Test
    fun `reconstructContent - orphan ref dropped sets fixed true`() {
        val root = note("r", "Root", containedNotes = listOf("missing"))
        val raw = mapOf("r" to root)
        val (_, fixed) = reconstructNoteContent(root, raw, childrenByParent(raw))
        assertTrue(fixed)
    }

    @Test
    fun `reconstructContent - stray child appended sets fixed true`() {
        val root = note("r", "Root", containedNotes = emptyList())
        val stray = note("s", "Stray", parentNoteId = "r")
        val raw = mapOf("r" to root, "s" to stray)
        val (result, fixed) = reconstructNoteContent(root, raw, childrenByParent(raw))
        assertEquals("Root\nStray", result.content)
        assertTrue(fixed)
    }

    @Test
    fun `reconstructContent - preserves non-content fields`() {
        val root = note("r", "Root", containedNotes = listOf("c1"))
            .copy(path = "my/path", state = "active")
        val c1 = note("c1", "Child", parentNoteId = "r")
        val raw = mapOf("r" to root, "c1" to c1)
        val (result, _) = reconstructNoteContent(root, raw, childrenByParent(raw))
        assertEquals("my/path", result.path)
        assertEquals("active", result.state)
    }

    // --- descendantIdsOf ---

    @Test
    fun `descendantIdsOf - returns empty set for note with no descendants`() {
        val rawNotes = mapOf("root" to note("root", "Root"))
        assertEquals(emptySet<String>(), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - returns IDs of non-deleted descendants`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root"),
            "c1" to note("c1", "Child 1", rootNoteId = "root"),
            "c2" to note("c2", "Child 2", rootNoteId = "root"),
        )
        assertEquals(setOf("c1", "c2"), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - excludes deleted descendants`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root"),
            "c1" to note("c1", "Alive", rootNoteId = "root"),
            "c2" to note("c2", "Deleted", rootNoteId = "root", state = "deleted"),
        )
        assertEquals(setOf("c1"), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - does not include notes from other roots`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root"),
            "c1" to note("c1", "My child", rootNoteId = "root"),
            "other" to note("other", "Other root"),
            "oc1" to note("oc1", "Other child", rootNoteId = "other"),
        )
        assertEquals(setOf("c1"), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - does not include the root note itself`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root", rootNoteId = null),
            "c1" to note("c1", "Child", rootNoteId = "root"),
        )
        assertEquals(setOf("c1"), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - includes deeply nested descendants`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root"),
            "a" to note("a", "Level 1", rootNoteId = "root", parentNoteId = "root"),
            "b" to note("b", "Level 2", rootNoteId = "root", parentNoteId = "a"),
            "c" to note("c", "Level 3", rootNoteId = "root", parentNoteId = "b"),
        )
        assertEquals(setOf("a", "b", "c"), descendantIdsOf("root", rawNotes))
    }

    @Test
    fun `descendantIdsOf - includeDeleted returns all descendants`() {
        val rawNotes = mapOf(
            "root" to note("root", "Root"),
            "c1" to note("c1", "Alive", rootNoteId = "root"),
            "c2" to note("c2", "Deleted", rootNoteId = "root", state = "deleted"),
        )
        assertEquals(
            setOf("c1", "c2"),
            descendantIdsOf("root", rawNotes, includeDeleted = true)
        )
    }
}

File: web/src/__tests__/data/NoteReconstruction.test.ts
import { describe, it, expect } from 'vitest'
import type { Note } from '../../data/Note'
import {
  rebuildAllNotes,
  rebuildAffectedNotes,
  reconstructNoteContent,
  reconstructNoteLines,
  indexChildrenByParent,
} from '../../data/NoteReconstruction'

function note(overrides: Partial<Note> & { id: string }): Note {
  return {
    userId: '',
    parentNoteId: null,
    content: '',
    createdAt: null,
    updatedAt: null,
    tags: [],
    containedNotes: [],
    state: null,
    path: '',
    rootNoteId: null,
    showCompleted: true,
    onceCache: {},
    ...overrides,
  }
}

function toMap(notes: Note[]): Map<string, Note> {
  const map = new Map<string, Note>()
  for (const n of notes) map.set(n.id, n)
  return map
}

// Re-export the production helper with a shorter name for test readability.
const childrenByParent = indexChildrenByParent

describe('rebuildAllNotes', () => {
  it('single root with no children returns as-is', () => {
    const root = note({ id: 'r1', content: 'Hello' })
    const result = rebuildAllNotes(toMap([root]))
    expect(result.notes).toHaveLength(1)
    expect(result.notes[0]).toBe(root)
    expect(result.notesNeedingFix.size).toBe(0)
  })

  it('root with descendants reconstructs content from tree', () => {
    const root = note({ id: 'r1', content: 'Title', containedNotes: ['c1', 'c2'] })
    const c1 = note({ id: 'c1', content: 'Line one', parentNoteId: 'r1', rootNoteId: 'r1' })
    const c2 = note({ id: 'c2', content: 'Line two', parentNoteId: 'r1', rootNoteId: 'r1' })
    const result = rebuildAllNotes(toMap([root, c1, c2]))
    expect(result.notes[0]!.content).toBe('Title\nLine one\nLine two')
    expect(result.notesNeedingFix.size).toBe(0)
  })

  it('orphan refs in containedNotes dropped and mark needsFix', () => {
    const root = note({ id: 'r1', content: 'Root', containedNotes: ['c1', 'c2', 'c3'] })
    const c2 = note({ id: 'c2', content: 'Only child', parentNoteId: 'r1' })
    const result = rebuildAllNotes(toMap([root, c2]))
    expect(result.notes[0]!.content).toBe('Root\nOnly child')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('stray child linked via parentNoteId appended at end and marks needsFix', () => {
    const root = note({ id: 'r1', content: 'Root', containedNotes: ['c1'] })
    const c1 = note({ id: 'c1', content: 'Known', parentNoteId: 'r1' })
    const c2 = note({ id: 'c2', content: 'Stray', parentNoteId: 'r1' })
    const result = rebuildAllNotes(toMap([root, c1, c2]))
    expect(result.notes[0]!.content).toBe('Root\nKnown\nStray')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('all declared refs orphaned but strays present reconstructs from strays', () => {
    const root = note({ id: 'r1', content: 'Alarms', containedNotes: ['orphan1', 'orphan2'] })
    const real1 = note({ id: 'real1', content: 'Real 1', parentNoteId: 'r1' })
    const real2 = note({ id: 'real2', content: 'Real 2', parentNoteId: 'r1' })
    const result = rebuildAllNotes(toMap([root, real1, real2]))
    expect(result.notes[0]!.content).toBe('Alarms\nReal 1\nReal 2')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('deleted child dropped and marks needsFix', () => {
    const root = note({ id: 'r1', content: 'Root', containedNotes: ['c1', 'c2'] })
    const c1 = note({ id: 'c1', content: 'Live', parentNoteId: 'r1' })
    const c2 = note({ id: 'c2', content: 'Deleted', parentNoteId: 'r1', state: 'deleted' })
    const result = rebuildAllNotes(toMap([root, c1, c2]))
    expect(result.notes[0]!.content).toBe('Root\nLive')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('child with mismatched parentNoteId dropped from ref list', () => {
    const root = note({ id: 'r1', content: 'Root', containedNotes: ['c1'] })
    const c1 = note({ id: 'c1', content: 'Not mine', parentNoteId: 'someone-else' })
    const result = rebuildAllNotes(toMap([root, c1]))
    expect(result.notes[0]!.content).toBe('Root')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('nested strays render at correct depth', () => {
    const root = note({ id: 'r1', content: 'Root', containedNotes: ['c1'] })
    const c1 = note({ id: 'c1', content: 'Known', parentNoteId: 'r1' })
    const grandchild = note({ id: 'gc1', content: 'Stray gc', parentNoteId: 'c1' })
    const result = rebuildAllNotes(toMap([root, c1, grandchild]))
    expect(result.notes[0]!.content).toBe('Root\nKnown\n\tStray gc')
    expect([...result.notesNeedingFix]).toEqual(['r1'])
  })

  it('multiple roots are each reconstructed independently', () => {
    const root1 = note({ id: 'r1', content: 'First' })
    const root2 = note({ id: 'r2', content: 'Second', containedNotes: ['c1'] })
    const child = note({ id: 'c1', content: 'Child', parentNoteId: 'r2' })
    const result = rebuildAllNotes(toMap([root1, root2, child]))
    expect(result.notes.find(n => n.id === 'r1')!.content).toBe('First')
    expect(result.notes.find(n => n.id === 'r2')!.content).toBe('Second\nChild')
  })

  it('orphaned child (rootNoteId points to non-existent root) is excluded from top-level', () => {
    const orphan = note({ id: 'o1', content: 'Orphan', parentNoteId: 'missing', rootNoteId: 'missing' })
    const result = rebuildAllNotes(toMap([orphan]))
    expect(result.notes).toHaveLength(0)
  })

  it('deleted root still returned as top-level', () => {
    const root = note({ id: 'r1', content: 'Deleted root', state: 'deleted' })
    const result = rebuildAllNotes(toMap([root]))
    expect(result.notes).toHaveLength(1)
  })
})

describe('rebuildAffectedNotes', () => {
  it('single root changed', () => {
    const existing = [note({ id: 'a', content: 'Old' }), note({ id: 'b', content: 'Beta' })]
    const rawNotes = toMap([note({ id: 'a', content: 'New' }), note({ id: 'b', content: 'Beta' })])
    const result = rebuildAffectedNotes(existing, new Set(['a']), rawNotes)
    expect(result.notes.find(n => n.id === 'a')!.content).toBe('New')
  })

  it('root deleted (removed from rawNotes)', () => {
    const existing = [note({ id: 'a' }), note({ id: 'b' })]
    const rawNotes = toMap([note({ id: 'b' })])
    const result = rebuildAffectedNotes(existing, new Set(['a']), rawNotes)
    expect(result.notes).toHaveLength(1)
    expect(result.notes[0]!.id).toBe('b')
  })

  it('new root added', () => {
    const existing = [note({ id: 'a' })]
    const rawNotes = toMap([note({ id: 'a' }), note({ id: 'b', content: 'new' })])
    const result = rebuildAffectedNotes(existing, new Set(['b']), rawNotes)
    expect(result.notes).toHaveLength(2)
  })

  it('unchanged roots preserved by reference', () => {
    const rootA = note({ id: 'a' })
    const rootB = note({ id: 'b' })
    const existing = [rootA, rootB]
    const rawNotes = toMap([note({ id: 'a', content: 'changed' }), rootB])
    const result = rebuildAffectedNotes(existing, new Set(['a']), rawNotes)
    expect(result.notes[1]).toBe(rootB)
  })

  it('returns same reference when no changes', () => {
    const rootA = note({ id: 'a' })
    const existing = [rootA]
    const rawNotes = toMap([note({ id: 'a' })])
    const result = rebuildAffectedNotes(existing, new Set(['x']), rawNotes)
    expect(result.notes).toBe(existing)
  })

  it('fix applied marks notesNeedingFix', () => {
    const existing = [note({ id: 'r', content: 'Old' })]
    const rawNotes = toMap([note({ id: 'r', content: 'New', containedNotes: ['orphan'] })])
    const result = rebuildAffectedNotes(existing, new Set(['r']), rawNotes)
    expect([...result.notesNeedingFix]).toEqual(['r'])
  })
})

describe('reconstructNoteContent', () => {
  it('no children or refs returns same instance', () => {
    const n = note({ id: 'a', content: 'Plain' })
    const raw = toMap([n])
    const [result, fixed] = reconstructNoteContent(n, raw, childrenByParent(raw))
    expect(result).toBe(n)
    expect(fixed).toBe(false)
  })

  it('declared children in order', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'c2'] })
    const c1 = note({ id: 'c1', content: 'A', parentNoteId: 'r' })
    const c2 = note({ id: 'c2', content: 'B', parentNoteId: 'r' })
    const raw = toMap([root, c1, c2])
    const [result, fixed] = reconstructNoteContent(root, raw, childrenByParent(raw))
    expect(result.content).toBe('Root\nA\nB')
    expect(fixed).toBe(false)
  })

  it('legacy "" in containedNotes is dropped as an orphan (post-migration)', () => {
    // Pre-migration, "" was a spacer sentinel. Post-migration, all empty
    // lines are real docs — an "" entry is data corruption that the orphan-
    // drop path catches and flags via `fixed = true`.
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', '', 'c2'] })
    const c1 = note({ id: 'c1', content: 'A', parentNoteId: 'r' })
    const c2 = note({ id: 'c2', content: 'B', parentNoteId: 'r' })
    const raw = toMap([root, c1, c2])
    const [result, fixed] = reconstructNoteContent(root, raw, childrenByParent(raw))
    expect(result.content).toBe('Root\nA\nB')
    expect(fixed).toBe(true)
  })

  it('orphan ref dropped sets fixed true', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['missing'] })
    const raw = toMap([root])
    const [, fixed] = reconstructNoteContent(root, raw, childrenByParent(raw))
    expect(fixed).toBe(true)
  })

  it('stray child appended sets fixed true', () => {
    const root = note({ id: 'r', content: 'Root' })
    const stray = note({ id: 's', content: 'Stray', parentNoteId: 'r' })
    const raw = toMap([root, stray])
    const [result, fixed] = reconstructNoteContent(root, raw, childrenByParent(raw))
    expect(result.content).toBe('Root\nStray')
    expect(fixed).toBe(true)
  })

  it('preserves non-content fields', () => {
    const root = note({
      id: 'r',
      content: 'Root',
      containedNotes: ['c1'],
      tags: ['important'],
      state: 'active',
      path: 'my-note',
    })
    const c1 = note({ id: 'c1', content: 'Child', parentNoteId: 'r' })
    const raw = toMap([root, c1])
    const [result] = reconstructNoteContent(root, raw, childrenByParent(raw))
    expect(result.tags).toEqual(['important'])
    expect(result.state).toBe('active')
    expect(result.path).toBe('my-note')
  })
})

describe('reconstructNoteLines', () => {
  it('returns per-line NoteLines with noteIds', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'c2'] })
    const c1 = note({ id: 'c1', content: 'First', parentNoteId: 'r' })
    const c2 = note({ id: 'c2', content: 'Second', parentNoteId: 'r' })
    const raw = toMap([root, c1, c2])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines).toEqual([
      { content: 'Root', noteId: 'r' },
      { content: 'First', noteId: 'c1' },
      { content: 'Second', noteId: 'c2' },
    ])
    expect(fixed).toBe(false)
  })

  it('drops declared child whose doc is missing from rawNotes (partial-sync window)', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'new_child', 'c2'] })
    const c1 = note({ id: 'c1', content: 'First', parentNoteId: 'r' })
    const c2 = note({ id: 'c2', content: 'Second', parentNoteId: 'r' })
    // new_child absent from raw map
    const raw = toMap([root, c1, c2])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'First', 'Second'])
    expect(fixed).toBe(true)
  })

  it('appends stray (parentNoteId but not in containedNotes) at end', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1'] })
    const c1 = note({ id: 'c1', content: 'Known', parentNoteId: 'r' })
    const stray = note({ id: 's', content: 'Stray', parentNoteId: 'r' })
    const raw = toMap([root, c1, stray])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'Known', 'Stray'])
    expect(lines.map(l => l.noteId)).toEqual(['r', 'c1', 's'])
    expect(fixed).toBe(true)
  })

  it('matches reconstructNoteContent output line-by-line (drift guard)', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'missing_c', 'c2'] })
    const c1 = note({ id: 'c1', content: 'First', containedNotes: ['gc1'], parentNoteId: 'r' })
    const c2 = note({ id: 'c2', content: 'Second', parentNoteId: 'r' })
    const gc1 = note({ id: 'gc1', content: 'GC', parentNoteId: 'c1' })
    const stray = note({ id: 's', content: 'Stray', parentNoteId: 'r' })
    const raw = toMap([root, c1, c2, gc1, stray])
    const cbp = childrenByParent(raw)

    const [lines, linesFixed] = reconstructNoteLines(root, raw, cbp)
    const [reconstructed, contentFixed] = reconstructNoteContent(root, raw, cbp)

    expect(lines.map(l => l.content).join('\n')).toBe(reconstructed.content)
    expect(linesFixed).toBe(contentFixed)
  })

  it('duplicate containedNotes ref placed once and flagged', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'c1'] })
    const c1 = note({ id: 'c1', content: 'Once', parentNoteId: 'r' })
    const raw = toMap([root, c1])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'Once'])
    expect(fixed).toBe(true)
  })

  it('cycle via containedNotes does not loop', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1'] })
    const c1 = note({ id: 'c1', content: 'First', containedNotes: ['r'], parentNoteId: 'r' })
    const raw = toMap([root, c1])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'First'])
    expect(fixed).toBe(true)
  })

  it('orphan ref and stray child in same walk both applied', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['c1', 'missing'] })
    const c1 = note({ id: 'c1', content: 'Declared', parentNoteId: 'r' })
    const stray = note({ id: 's', content: 'Stray', parentNoteId: 'r' })
    const raw = toMap([root, c1, stray])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'Declared', 'Stray'])
    expect(fixed).toBe(true)
  })

  it('deeply nested stray renders at correct depth', () => {
    const root = note({ id: 'r', content: 'Root', containedNotes: ['a'] })
    const a = note({ id: 'a', content: 'A', containedNotes: ['b'], parentNoteId: 'r' })
    const b = note({ id: 'b', content: 'B', containedNotes: ['c'], parentNoteId: 'a' })
    const c = note({ id: 'c', content: 'C', parentNoteId: 'b' })
    const stray = note({ id: 's', content: 'Stray', parentNoteId: 'c' })
    const raw = toMap([root, a, b, c, stray])
    const [lines, fixed] = reconstructNoteLines(root, raw, childrenByParent(raw))
    expect(lines.map(l => l.content)).toEqual(['Root', 'A', '\tB', '\t\tC', '\t\t\tStray'])
    expect(fixed).toBe(true)
  })
})
